### PR TITLE
Error handling empty message and unknown fields

### PR DESCRIPTION
**Pull request by fzrvic**

### CODE REVIEW

File: src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ProtoUtil {
+    public static boolean isUnknownFieldExist(DynamicMessage dynamicMessage) {
+        if (dynamicMessage == null) {
+            return false;
+        }
+        List<DynamicMessage> dynamicMessageFields = new LinkedList<>();
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java:14]: missing the root message

File: src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java [hunk 2]
```diff
@@ -41,8 +43,15 @@ public ObjectStorageSink(Instrumentation instrumentation, String sinkType, Write
             try {
                 Record record = messageDeSerializer.deSerialize(message);
                 offsetManager.addOffsetToBatch(writerOrchestrator.write(record), message);
+            } catch (EmptyMessageException e) {
+                message.setErrorInfo(new ErrorInfo(e, ErrorType.INVALID_MESSAGE_ERROR));
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java:47]: log the errors

File: src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageDeSerializer.java [hunk 3]
```diff
@@ -16,15 +20,24 @@
 
     public Record deSerialize(Message message) throws DeserializerException {
         try {
-            DynamicMessage dynamicMessage = protoParser.parse(message.getLogMessage());
+            if (message.getLogMessage() == null || message.getLogMessage().length == 0) {
+                throw new EmptyMessageException();
+            }
+
+            DynamicMessage logMessage = protoParser.parse(message.getLogMessage());
```
Comment (sravankorumilli) [inline src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageDeSerializer.java:27]: name of the field use dynamicMessage, logmessage is before deserialization

File: src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java (hunk 2 above)
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/sink/objectstorage/ObjectStorageSink.java:47]: will add log

File: src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageDeSerializer.java (hunk 3 above)
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/sink/objectstorage/message/MessageDeSerializer.java:27]: yes, will rename

File: src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java (hunk 1 above)
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java:14]: similar with beast changes, every node of DynamicMessage will be added automatically when traversing the DynamicMessage trees, so there is no need to add the root explicitly on accumulator

File: src/main/java/io/odpf/firehose/sink/bigquery/converter/MessageRecordConverter.java [hunk 4]
```diff
@@ -27,39 +34,49 @@
...
-            validRecords.add(new Record(message, columns));
         }
         return new Records(validRecords, invalidRecords);
     }
 
-    private Map<String, Object> mapToColumns(Message message) throws InvalidProtocolBufferException {
-        Map<String, Object> columns = Collections.emptyMap();
+    private Record createRecord(Instant now, Message message) throws DeserializerException {
+        if (message.getLogMessage() == null || message.getLogMessage().length == 0) {
+            log.info("empty message found at offset: {}, partition: {}", message.getOffset(), message.getPartition());
```
[the remark this answers is not in the document]
Comment (fzrvic) [reply src/main/java/io/odpf/firehose/sink/bigquery/converter/MessageRecordConverter.java:60]: will change it to warns/errors

File: src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java (hunk 1 above)
Comment (sravankorumilli) [reply src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java:14]: missed it.

File: src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+public class ProtoUtil {
+    public static boolean isUnknownFieldExist(DynamicMessage root) {
```
Comment (lavkesh) [inline src/main/java/io/odpf/firehose/sink/common/ProtoUtil.java:13]: rename:
 isUnknownFieldExist -> hasUnknownFields.
 ProtoUtil -> ProtoUtils

**Pull request merged**